### PR TITLE
drivers: optee: Introduce basic OP-TEE driver to Zephyr 

### PR DESCRIPTION
**Pull request by oleksiimoisieiev**
This introduces the TEE subsystem definition and OP-TEE driver basic implementation with the unit tests
This is preliminary implementation which implements basic syscalls, needed to communicate wit OP-TEE OS.
Supplicant support should be implemented in future and probably posted in the separate PR, or merged into this one if it will be still open.

### CODE REVIEW

File: include/zephyr/drivers/tee.h [hunk 1]
```diff
@@ -0,0 +1,760 @@
...
+#define TEEC_ERROR_EXCESS_DATA			0xFFFF0004
+#define TEEC_ERROR_BAD_FORMAT			0xFFFF0005
+#define TEEC_ERROR_BAD_PARAMETERS		0xFFFF0006
+#define TEEC_ERROR_BAD_STATE			0xFFFF0007
+#define TEEC_ERROR_ITEM_NOT_FOUND		0xFFFF0008
+#define TEEC_ERROR_NOT_IMPLEMENTED		0xFFFF0009
+#define TEEC_ERROR_NOT_SUPPORTED		0xFFFF000A
+#define TEEC_ERROR_NO_DATA			0xFFFF000B
+#define TEEC_ERROR_OUT_OF_MEMORY		0xFFFF000C
+#define TEEC_ERROR_BUSY			0xFFFF000D
```
Comment (firscity) [inline include/zephyr/drivers/tee.h]: Please, recheck formatting in this file

File: include/zephyr/drivers/tee.h [hunk 2]
```diff
@@ -0,0 +1,760 @@
...
+	const struct tee_driver_api *api = (const struct tee_driver_api *)dev->api;
+
+	if (!api->invoke_func) {
+		return -ENOSYS;
+	}
+
+	return api->invoke_func(dev, arg, num_param, param);
+}
+
+static inline int tee_add_shm(const struct device *dev, void *addr, size_t align, size_t size,
```
Comment (firscity) [inline include/zephyr/drivers/tee.h]: Regarding conversation in PR for zephyr-v3.3.0-xt: I don't see problem to add 1 more source file to get rid of this inlines.

File: tests/drivers/tee/tee/boards/native_posix_64.overlay [hunk 3]
```diff
@@ -0,0 +1,12 @@
+/* Copyright (c) 2023 EPAM Systems
```
Comment (firscity) [inline tests/drivers/tee/tee/boards/native_posix_64.overlay]: Newline at 1st line and before root node

File: tests/drivers/tee/optee/boards/native_posix_64.overlay [hunk 4]
```diff
@@ -0,0 +1,12 @@
+/* Copyright (c) 2023 EPAM Systems
```
Comment (firscity) [inline tests/drivers/tee/optee/boards/native_posix_64.overlay]: Same for Copyright formatting

File: tests/drivers/tee/optee/src/main.c [hunk 5]
```diff
@@ -0,0 +1,311 @@
+/*
+ * Copyright (c) 2023 Epam Systems
```
Comment (firscity) [inline tests/drivers/tee/optee/src/main.c]: EPAM

File: include/zephyr/drivers/tee.h (hunk 1 above)
Comment (oleksiimoisieiev) [reply include/zephyr/drivers/tee.h]: +

File: include/zephyr/drivers/tee.h (hunk 2 above)
Comment (oleksiimoisieiev) [reply include/zephyr/drivers/tee.h]: fixed

File: tests/drivers/tee/tee/boards/native_posix_64.overlay (hunk 3 above)
Comment (oleksiimoisieiev) [reply tests/drivers/tee/tee/boards/native_posix_64.overlay]: +

File: tests/drivers/tee/optee/boards/native_posix_64.overlay (hunk 4 above)
Comment (oleksiimoisieiev) [reply tests/drivers/tee/optee/boards/native_posix_64.overlay]: +

File: tests/drivers/tee/optee/src/main.c (hunk 5 above)
Comment (oleksiimoisieiev) [reply tests/drivers/tee/optee/src/main.c]: +

File: drivers/tee/optee/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2023 Epam Systems
```
Comment (firscity) [inline drivers/tee/optee/CMakeLists.txt]: EPAM

File: drivers/tee/optee/optee.c [hunk 7]
```diff
@@ -0,0 +1,933 @@
+/*
+ * Copyright 2023 Epam Systems
```
Comment (firscity) [inline drivers/tee/optee/optee.c]: EPAM

File: include/zephyr/drivers/tee.h [hunk 8]
```diff
@@ -0,0 +1,694 @@
...
+ *                              Application process is provided.
+ * TEEC_LOGIN_APPLICATION       Login data about the running Client Application
+ *                              itself is provided.
+ * TEEC_LOGIN_USER_APPLICATION  Login data about the user and the running
+ *                              Client Application itself is provided.
+ * TEEC_LOGIN_GROUP_APPLICATION Login data about the group and the running
+ *                              Client Application itself is provided.
+ */
+#define TEEC_LOGIN_PUBLIC		0x00000000
+#define TEEC_LOGIN_USER		0x00000001
```
Comment (firscity) [inline include/zephyr/drivers/tee.h]: :)

**Review by firscity — APPROVED** [review]
with last comments fixed
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`

**Pull request merged**